### PR TITLE
Chapter4 solns

### PR DESCRIPTION
**Pull request by sushrut111**
### Solutions for Chapter {NUMBER}

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -305,7 +305,12 @@ typeclasses for standard data types.
 -}
 data List a
     = Empty
-    | Cons a (List a)
+    | Cons a (List a) deriving(Show)
+
+instance Functor List where
+    fmap :: (a -> b) -> List a -> List b
+    fmap _ Empty = Empty
+    fmap f (Cons a b) = Cons (f a) (fmap f b)
```
Comment (vrom911) [inline src/Chapter4.hs:313]: Wonderful!  Great recursive implementation 👍🏼

File: src/Chapter4.hs [hunk 2]
```diff
@@ -472,10 +477,12 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Trap x) _ = (Trap x)
+    (<*>) (Reward f) x = fmap f x
```
Comment (vrom911) [inline src/Chapter4.hs:484]: Great usage of the `Functor` instance of `Secret` 🗺️ 🔝

File: src/Chapter4.hs [hunk 3]
```diff
@@ -488,7 +495,26 @@ Implement the 'Applicative' instance for our 'List' type.
...
+{-
+data List a
+    = Empty
+    | Cons a (List a) deriving(Show)
+
+-}
 
+append :: List a -> List a -> List a
+append Empty l = l
+append l Empty = l
```
Comment (vrom911) [inline src/Chapter4.hs:507]: Nice catch! This makes implementation way more efficient 🚤

File: src/Chapter4.hs [hunk 4]
```diff
@@ -628,8 +662,11 @@ Can you implement a monad version of AND, polymorphic over any monad?
 
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
+
+
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM a b = (fmap (\x-> (&&) x) a) <*> b
```
Comment (vrom911) [inline src/Chapter4.hs:668]: The ordinary `&&` if lazy on the second argument. Meaning that if it meets `False` as the first argument, there is no need to calculate another arg. In `andM` we would like to keep the same behaviour, so the 
```
Just False `andM` _
``` 
should be `Just False` 🙂 

With the `Applicative` constraint, it is not possible to make it lazy. That is why we suggest using `Monad` 🙂 
But in reality, you only use `Applicative in here 🙂

File: src/Chapter4.hs [hunk 5]
```diff
@@ -674,6 +711,79 @@ Specifically,
 -}
 
 
+data Btree a = Btree
+  {
+    lchild :: Btree a,
+    rchild :: Btree a,
+    value :: a
+  } | Null deriving (Show)
```
Comment (vrom911) [inline src/Chapter4.hs:719]: That is right representation of the Tree.

However, having record fields inside the sum type considered antipattern and highly discouraged.
As we discussed in previous Chapter, each record name is the function in reality, but as it is the sum type, these functions are partial (can fail in runtime). So it is better to not use that.

In this case you have two options. First, and most common, to use unnamed fields:
```haskell
data Btree a 
    = Btree (Btree a) (BTree a) a 
    | Null 
    deriving (Show)
```

Another option, to use additional record type, and use it unnamed inside the tree

```haskell
data Node a = Node
    { lchild :: Btree a
    , rchild :: Btree a
    , value :: a
    }

data Btree a 
    = Tree (Node a) 
    | Null 
    deriving (Show)
```

File: src/Chapter4.hs [hunk 6]
```diff
@@ -674,6 +711,79 @@ Specifically,
...
+
+instance Functor Btree where
+    fmap :: (a -> b) -> Btree a -> Btree b
+    fmap _ Null = Null
+    fmap f (Btree a b c) = Btree (fmap f a) (fmap f b) (f c)
+
+reverseTree :: Btree a -> Btree a
+reverseTree Null = Null
+reverseTree (Btree a b c) = Btree (reverseTree b) (reverseTree a) c
```
Comment (vrom911) [inline src/Chapter4.hs:728]: Yep, it is fascinating, how easy it is to reverse the Tree! ⭐

**Review by vrom911 — APPROVED** [review]
Wohoo! All chapters are ready!
You nailed this complicated one, this is really cool! I enjoyed your solutions, and great job with Trees 🌴 This is not an easy one! 🔝 

Thanks for taking this course. Hope you had fun and found something interesting for yourself in it ❤️ 
If you have any questions or want to discuss some topics with us, don't hesitate to reach out!
Also, We appreciate any feedback! 🙏🏼

File: src/Chapter4.hs (hunk 5 above)
Comment (sushrut111) [reply src/Chapter4.hs:719]: This is very helpful, thanks!

File: src/Chapter4.hs (hunk 4 above)
Comment (sushrut111) [reply src/Chapter4.hs:668]: Noted. Thank you!

**Comment by sushrut111** [comment]
> Wohoo! All chapters are ready!
> You nailed this complicated one, this is really cool! I enjoyed your solutions, and great job with Trees 🌴 This is not an easy one! 🔝
> 
> Thanks for taking this course. Hope you had fun and found something interesting for yourself in it ❤️
> If you have any questions or want to discuss some topics with us, don't hesitate to reach out!
> Also, We appreciate any feedback! 🙏🏼

I had so much fun in this journey with you guys! I appreciate your efforts to help everyone learn haskell.

**Pull request merged**